### PR TITLE
Add Provider shutdown command

### PR DESCRIPTION
**Pull request by 7hunderbird**
This adds the command which paves the way for a `provider` encapsulation we can re-use.

Essentially we want to make a command that is convenient and safe to terminate the servers currently running on EC2.

This is a first cut, we can add layers of control to it.  The "meat" of the search is:

```
aws ec2 describe-instances --filters "Name=instance-state-name,Values=running" | jq -c -r '.Reservations[] .Instances[] .InstanceId' > instance_list
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,8 +1,9 @@
+instance_list
```
Comment (jahio) [inline .gitignore]: Suggestion:

```
/instance-list*
```

Limits scope to repo root directory only and allows us to append things like a datestamp, uuid, whatever to the filename for preservation of output relative to a specific point in time (see addl. comments below)

File: bin/nbb [hunk 2]
```diff
@@ -344,6 +339,24 @@ find_centos_7_ami_ids() {
   done
 }
 
+shutdown_instances() {
+  aws ec2 describe-instances --filters "Name=instance-state-name,Values=running" | jq -c -r '.Reservations[] .Instances[] .InstanceId' > instance_list
```
Comment (jahio) [inline bin/nbb]: Suggestion: invert the state name if possible to effectively say: "show me all instances that are NOT in state `terminated`". Catches any state (terminating, shutting-down, stopped, running, error, etc.) that could potentially lead to dependency violations.

**New dependency in `$PATH`:** `jq`. Remember to list it in the readme under what you very cleverly named "the shopping list" for anyone not using `brew`.

Suggestion: instead of just throwing it all into one file with a hard-coded name, it may be useful to put the instance list in a file with a date-stamped suffix, i.e. `instance-list.2016-06-03-16-55-33UTC.txt`. (_This is why I suggested tweaking the .gitignore pattern above_). With this format we can let the user decide to keep output state as reported at a given date/time. Can provide a "point in time" state snapshot that might be useful to them for unknown reasons, and will almost certainly be useful _for us_ when debugging.

Suggestion: isolate this line into a separate shell command called  `status` or something similar. Since we're just getting a list of instances running, we can pull that out into its own shell command and decouple the "discovery" responsibilty from the "destroy" responsibility. That enables us to create a `make provider-status` command (or equivalent) that could be helpful in debugging as well.

File: bin/nbb [hunk 3]
```diff
@@ -344,6 +339,24 @@ find_centos_7_ami_ids() {
...
+  echo "You are about to terminate the following instances:"
+  cat instance_list
+  read -r -p "Are you sure? [y/N] " response
```
Comment (jahio) [inline bin/nbb:357]: Suggestion: allow the user to pass in an option like `--yes-im-sure`; if _and only if_ that option is present (e.g. it's **not** the default), don't bother asking. Allows for semi-scriptable interface and/or some form of automation. Might not be advisable or obvious how it's useful but I always come at these things from the perspective of "_everything_ should be able to run non-interactively, no matter what, even if it's dangerous - as long as you force the user to opt-in to the dangerous stuff first".

File: bin/nbb [hunk 4]
```diff
@@ -344,6 +339,24 @@ find_centos_7_ami_ids() {
...
+  if [[ $response =~ ^([yY][eE][sS]|[yY])$ ]]
+  then
+    IFS=$'\n'  # make newlines the separator
+    for i in `cat ./instance_list`; do
+      echo "Shutting down $i"
+      aws ec2 terminate-instances --instance-ids $i
```
Comment (jahio) [inline bin/nbb:363]: Suggestion: perhaps wrap the `aws ec2` call in a subshell function and/or background the process (maybe make note of its PID somewhere too?). **I'm worried that if AWS is slow, it could take hours upon hours to delete all your instances**, especially if you have a lot of them. In other words, I don't know if this command is syncrounous/blocking or not, and either way we can't gaurantee it'll _stay_ that way in future releases.

File: bin/nbb [hunk 5]
```diff
@@ -344,6 +339,24 @@ find_centos_7_ami_ids() {
...
+    done
+  else
+    echo "Nothing shutdown, exiting..."
+  fi
+  unset IFS
```
Comment (jahio) [inline bin/nbb:368]: Suggestion: if the user passes in an argument like `--cleanup` when calling this command/function/wrapper, _automatically remove the generated instances list file for them_ right here. Lets users keep their filesystem tidy if they don't care about having that point in time snapshot of state I mentioned above, and allows other users to keep that data if they so desire; all that's required is for the user to tell us how they want it. :smiley:

File: bin/nbb (hunk 4 above)
Comment (sethlindberg) [inline bin/nbb:363]: In my experience, this isn't an issue.  AWS shuts things down relatively quickly except in extreme edge cases.  The only situation where this is different is in instance termination, where it shuts things down right away and then gives a four hour grace period to catch a possible mistake or bad actor.

In general in the state AWS is right now, it's very fault-tolerant.  Wasn't always this way, but it is now.

Comment (jahio) [inline bin/nbb:363]: @sethlindberg I'm sure they're relatively quick in most cases as you pointed out; my concern is more around questions like, "what if I have 100 instances to terminate!?" 😧 

If this is _blocking_, we'll find the user stuck in that loop for quite some time.

Now, _is that actually bad?_ In this case there might be a good reason to go fully synchronous here; if we can't do anything else with our teardown until all these instances are gone anyway, maybe we just have to bite the bullet and wait on them one-by-one. However, this creates another problem: if somehow one of these calls times out and the `aws` command exits status `1`, _will it force an `exit 1` of the entire orchestration script, all the way up the stack?_

> The only situation where this is different is in instance termination, where it shuts things down right away and then gives a four hour grace period to catch a possible mistake or bad actor.

I wanted to point this out actually; the use case here is to prevent `make destroy` from barfing up at us due to AWS dependency violations. One thing I've seen pretty often is that an instance still exists inside a subnet so the entire `make destroy` process explodes and you don't know where you stand anymore. It's rather frustrating indeed.

The problem is that we _DO_ need to **terminate** these instances before we can be sure not to get thrown said `DepdendencyViolation` errors. AFAIK a simple change of state to `shutdown` isn't enough to do it. I've seen instances in that state in my test account derail `make destroy` in the past (plus I think the user's still billed for a `shutdown` instance, right?)

Anyway, I wanted to point out the consequences of making this synchronous. If I'm in the minority opinion here, hey no worries! :smiley: I just wanted to be sure it was at least discussed in some fashion before glossing it over, that's all.

Comment (sethlindberg) [inline bin/nbb:363]: If that's the problem @7hunderbird is trying to solve, then this is entirely the wrong way to go about it, in my understanding.

The old way logs into CF, asks the bosh master to delete all the cloud foundry instances (runner/0, etc.)  -- which used to work and stopped working  Why doesn't seem to have been investigated.

This way uses a command line to delete three instances ( `--filters "Name=tag:Name,Values=bastion,nat,bosh/0"` ) two of which are _managed by terraform_ and doesn't cover any of the cloud foundry instances.  There will still be instances that will stick around if someone deploys cloudfoundry to them.

Comment (jahio) [inline bin/nbb:363]: I think you might not have been present for the conversations we've had about this over the last couple weeks (just catching you up, not an accusatory statement!). Basically it boils down to this: `make destroy` calls Terraform's destroy command which is **supposed** to cascade down, _deleting all the things inside the VPCs, subnets, etc. automatically_.

The problem is that while `terraform` seems to be working fine, **Amazon** isn't behaving according to its own API. That cascade effect winds up failing roughly half the time because Amazon's API throws Dependency Violation errors, meaning that it refuses to delete an object because something still exists that depends on it. I've personally seen this at least a dozen times just myself; I'm sure the rest of the team has too.

So basically it boils down to the fact that we need to be able to see, _outside of `terraform`_, what's running (those "dependent" objects) so we can be sure to terminate them properly. **THEN** `terraform destroy` can actually work because the Amazon API won't barf on us.

💩 

_Edit: Clarification_ The cascading deletion thing I mentioned, that's how Amazon's object deletion is supposed to work. If you have an instance inside a subnet let's say, and you delete the subnet, it's supposed to delete the instance _for_ you automatically **before** trying to delete said subnet. In reality, though, _Amazon's own API_ is trying to delete the "containing" thing before all its "children" have been removed.

File: bin/nbb (hunk 5 above)
Comment (jahio) [inline bin/nbb:368]: RE: `--cleanup`: not necessary for `1.0` or initial release so let's just ignore this for now and let it serve as a future feature idea.

File: bin/nbb (hunk 4 above)
Comment (sethlindberg) [inline bin/nbb:363]: Er, no. I added this to the trello card last week, but it wasn't read.

You can see what `make destroy` does before this patch.  It logs into the bastion host, runs a provision destroy (which has stopped working for some reason) and then calls terraform to destroy the rest.

Terraform only destroys stuff that it manages.  Since it does't manage bosh (that's handled by bosh-init) or any of the stuff in cloud foundry (that gets managed by bosh during the deploy) it doesn't know to destroy them.

This is a case of having two different configurations for AWS stuff - the terraform component for the VPC, and the bosh component.  The bosh component isn't working correctly, which causes the terraform component to not work correctly (since there are instances that terraform doesn't manage on VPC subnets that it does manage)

Comment (sethlindberg) [inline bin/nbb:363]: _specifically_, the file that gets called on the bastion host during a `make destroy` is put there during the `make provision-base` and is called here:  https://github.com/cloudfoundry-community/aws-nat-bastion-bosh-cf/blob/master/scripts/provision-bastion#L503 

After this is done (in theory) and whether or not that command succeeds, `make destroy` immediately calls terraform's destroy, which will fail if there are instances that it does not manage in the vpc that it does manage, but possibly not after it has destroyed the bastion host that it does manage.

Comment (7hunderbird) [inline bin/nbb:363]: @sethlindberg @jaustinhughey I'm going to continue this as a WIP branch, essentially though the idea is to have `make destroy` only be concerned with terraform teardown.  And `make provision-shutdown` (which is what I'll be renaming it to soon) be concerned with shutting down what gets started up post-terraform.  My POC was just can I get the `bosh/0` server to shut down.  Eventually I want to have a configurable list that is only the `bosh/0` + `cf` instances so that the `make provision-shutdown` only tearsdown the instances it's concerned with.

We will then make it clear in code and documentation that `make provision-shutdown` needs to be called before `make destroy`.

Comment (jahio) [inline bin/nbb:363]: > We will then make it clear in code and documentation that make provision-shutdown needs to be called before make destroy.

What do you think about combining those two together into their own `make` task? Perhaps `make cleanup` or something? `make nuclear-bombardment` `make EVERYTHINGOBOOM` :trollface:
